### DISCUSSION TITLE
Feature request: simple JSON view

### DISCUSSION THREAD

**Discussion by jamescridland** [skymethod/op3#18] [Ideas]
Hello!

I'm looking for a JSON output that would return four numbers, with four pieces of associated metadata, in a JSON object:

         ('seven-day-downloads'),
         ('seven-day-downloads-asof'),
         ('thirty-day-downloads'),
         ('thirty-day-downloads-asof'),
         ('downloads-count'),
         ('downloads-period'),
         ('audience-count'),
         ('audience-period')

I guess either this could be via an API call, and be a new portion of /worker/routes/api_queries.ts - or it could be just an alternate view of the main show page, in which case the URL could be as simple as /show/55f8007aec094899a02fd44273aa6558/headline.json (with perhaps a last-built date field).

This would enable a quick method of retrieving this headline data, which is lacking (and, selfishly, would allow me to retire a chunk of analytics code that I'm currently running myself).

It strikes me that if this were an alternate view of the main show page, it needn't use an API key and be simpler to produce; and given it's not private data in any way, shouldn't really need an API key to view this information. But I don't know which you'd prefer, given your infrastructure. Which would work better, and is there a way I can help with that?

**Comment by johnspurlock-skymethod** [comment]
Thanks for the suggestion James, I'll take a look.  Probably should be precomputed during the daily stats pass and returned via an api subendpoint for the show, something like `/api/1/shows/<show-id>/headline-stats`

If you are in a pinch for time, you can grab the stats for the show and compute it yourself (this is what the show stats pages do) using the undocumented show stats endpoint.  It's not documented yet since it may change a bit before I'm ready to commit to the api in the public docs.

e.g. show stats for podnews daily
https://op3.dev/api/1/shows/55f8007aec094899a02fd44273aa6558/stats?token=preview07ce

There are a few optional params if you need to go back in time more than a few months, see the searchParams parsing in [the method implementation](https://github.com/skymethod/op3/blob/9959c54c0057e67e4e887d68631f3bfd7651d6ad/worker/routes/api_shows.ts#L94) for what they do.

**Comment by jamescridland** [comment]
Thank you. That's quite the data dump! Goodness.

I'd be very happy to help with the headline stats API call if I knew how. My suspicion is that this could be a very cheap API call, and could even be used in directories.

**Comment by daveajones** [comment]
Once you settle on something I’d like to interface with these stats for PI as well.  Good call @jamescridland

**Comment by johnspurlock-skymethod** [comment]
@jamescridland just to clarify, you're looking for those 4 existing show-level metrics (and associated time ranges) at the top of each OP3 show stats page?

I've wanted a way to compute an episode-level 30-day download number for a while, since this is the number most folks look at for $$$ purposes, and doesn't bias based on publishing frequency.   This number is a bit trickier to compute, since each episode has different numbers (maybe a mean+variance would be better, or simply the latest episode that has reached 30 days).  

Anyway if I'm running over each show to precompute those existing headline box stats, perhaps I could use this as an excuse to start computing this number as well.

**Reply by jamescridland** [reply]
@johnspurlock-skymethod Yes, I'm looking for those 4 existing (actually, I don't care about the 7 day number). To clarify further, what I'd personally like is "downloads in the last calendar month" - then this is data that can be healthily cached.

Of course, a 30-day download number for episodes might be nice (or, indeed, any other data points - personally, I'd find "25% Apple // 20% Spotify // 35% Other" really interesting).

**Reply by johnspurlock-skymethod** [reply]
> "downloads in the last calendar month"

meaning downloads in the most recent completed calendar month?  ie would return downloads for Aug 1 to Aug 31 (inclusive) on any given day in September

**Reply by jamescridland** [reply]
Yes please.

(My assumption is that from 2 Sep 00:00 GMT, given a day for calculation, I could invalidate a cache and see August's numbers.)

**Reply by johnspurlock-skymethod** [reply]
Ok, from what I'm hearing, how about we start small:
```json
{
  "lastCalendarMonthDownloads": 123123,
  "lastCalendarMonthAudience": 12345,
  "lastCalendarMonth": "yyyy-mm"
}
```

Callers will have to deal with normalizing months with different numbers of days if doing month-to-month comparisons

**Reply by johnspurlock-skymethod** [reply]
Ok first pass is up:  `https://op3.dev/api/1/shows/<show-id>/summary-stats?token=<token>`

**Reply by jamescridland** [reply]
"show-id" is the Guid?

So for https://op3.dev/show/396d9ae0-da7e-5557-b894-b606231fa3ea the summary stats are at

https://op3.dev/api/1/shows/396d9ae0-da7e-5557-b894-b606231fa3ea/summary-stats?token=preview07ce

?

**Reply by johnspurlock-skymethod** [reply]
It's the immutable OP3 show uuid, you can look it by podcast guid with: https://op3.dev/api/docs#tag/shows/operation/viewShowInformation
